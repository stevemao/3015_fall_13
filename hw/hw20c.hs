module Tree where

import Data.Functor
import Control.Applicative
import qualified Data.Foldable  as F
import Data.Monoid


data Tree a = Leaf | Node (Tree a) a (Tree a)  deriving (Eq,Show)

mktree [] = Leaf
mktree xs = Node (mktree ls) r (mktree rs)
  where (ls,r:rs) = splitAt (length xs `div` 2) xs

-- instance Functor Tree where
  -- fill this in


-- instance Applicative Tree where
--   pure x = 
--   Leaf <*> _ = 
--   (Node l f r) <*> Leaf = 
--   (Node l f r) <*> (Node l1 x r1)  = 


-- the following implementation if right out of LYAHFGG -- but you should read it anyway.

instance F.Foldable Tree where
  foldMap f Leaf = mempty
  foldMap f (Node l x r) =  F.foldMap f l `mappend` f x `mappend` F.foldMap f r

flatten :: Tree a -> [a]
flatten = F.foldr (\x xs -> x:xs) []
