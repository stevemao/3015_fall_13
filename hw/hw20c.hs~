module Tree where

import Data.Functor
import Control.Applicative
import qualified Data.Foldable  as F
import Data.Monoid


data Tree a = Leaf | Node (Tree a) a (Tree a)  deriving (Eq,Show)

mktree [] = Leaf
mktree xs = Node (mktree ls) r (mktree rs)
  where (ls,r:rs) = splitAt (length xs `div` 2) xs

-- instance Functor Tree where
  -- fill this in

-- instance Applicative Tree where
--   pure x = 
--   Leaf <*> _ = 
--   (Node l f r) <*> Leaf = 
--   (Node l f r) <*> (Node l1 x r1)  = 


-- instance F.Foldable Tree where
-- fill this in (see LYAHFGG)

flatten :: Tree a -> [a]
flatten = F.foldr (\x xs -> x:xs) []