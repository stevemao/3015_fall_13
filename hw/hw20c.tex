\documentclass[11pt]{article}

\include{header}
\newcommand{\quot}[1]{"{\tt{{#1}}}"}
\newcommand{\mybs}{$\backslash$}

\begin{document}
\homework{20}{15 November}

We have been discussing the type classes Functor, Applicative and
Monoid.  You can read about Foldable in Chapter 11 of LYAHFGG.

In this assignment you must instantiate a tree datatype as an instance
of the Functor type class and as an instance of the Applicative type
class.  

\begin{program**}
\> data Tree a = Leaf $\mid$ Node (Tree a) a (Tree a)  deriving (Eq,Show)\\
\end{program**}

Mapping over a tree is straightforward so the instantiation of Tree as an
instance of Functor is easy.  Your implementation of Tree as an instance of the
Applicative type class behave more like the {\it{ZipList}} instantiation as an
instance of Applicative than the standard list instantiation does. In other
words, the Applicative operator $\langle{}*\rangle$ (which for trees will be of
type {\it{Tree (a $\rightarrow$ b) $\rightarrow$ Tree a $\rightarrow$ Tree b}})
should apply nodes pairwise. If $t_1$ and $t_2$ are trees, $t_1 \langle*\rangle
t_2$ should apply the function in the root node of $t_1$ to the value stored in
the root node of $t_2$ and the function stored in the root of the left subtree
of $t_1$ should be applied to the value stored in the left subtree of $t_2$ and
so on.  Like the {\it{zip}} function, if either tree runs out of values (gets
to a Leaf), the resulting value should just be {\it{Leaf}}.  Here are some
small example behaviors.

\begin{program**}
\> *Tree$\rangle$ (Node Leaf (+7) Leaf) $\langle{}*\rangle$ Leaf \\
\> Leaf \\
\> *Tree$\rangle$ Leaf $\langle{}*\rangle$ (Node Leaf 1 Leaf) \\
\> Leaf \\
\> *Tree$\rangle$ (Node Leaf (+7) Leaf) $\langle{}*\rangle$ (Node Leaf 7 Leaf) \\
\> Node Leaf 14 Leaf \\
\> *Tree$\rangle$ (Node Leaf (+7) Leaf) $\langle{}*\rangle$ (Node Leaf 7 (Node Leaf 12 Leaf)) \\
\> Node Leaf 14 Leaf \\
\> *Tree$\rangle$ (Node Leaf (+7) Leaf) $\langle{}*\rangle$ (Node (Node Leaf 12 Leaf) 7 Leaf) \\
\> Node Leaf 14 Leaf \\
\> *Tree$\rangle$  \\
\end{program**}

As noted in the discussion of {\it{ZipList}} as an instance of Applicative
(read LYAHFGG) this means {\it{pure :: a $\rightarrow$ Tree a}} must result in
an infinite tree. So {\it{pure x}} is an infinite tree of nodes containing the
value $x$.  HINT: Lipova\u{c}a used {\em{repeat}} in his definition of
{\it{pure}} for {\it{ZipList}} but you could have also done it by a recursive
call to {\it{pure}} in its own defintion.  

\end{document}


% Local Variables:
% mode:latex
% comment-column:0
% comment-start: "% "
% compile-command: "pdflatex hw20c"
% fill-column:79
% End:




