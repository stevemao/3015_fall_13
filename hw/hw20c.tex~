\documentclass[11pt]{article}

\include{header}
\newcommand{\quot}[1]{"{\tt{{#1}}}"}
\newcommand{\mybs}{$\backslash$}

\begin{document}
\homework{20}{15 November}



\end{document}


% Local Variables:
% mode:latex
% comment-column:0
% comment-start: "% "
% compile-command: "pdflatex hw20c"
% fill-column:79
% End:




