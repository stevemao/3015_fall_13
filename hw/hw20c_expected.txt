*Tree> mktree []
mktree []
Leaf
*Tree> mktree [1]
mktree [1]
Node Leaf 1 Leaf
*Tree> mktree [1,2]
mktree [1,2]
Node (Node Leaf 1 Leaf) 2 Leaf
*Tree> mktree [1,2,3]
mktree [1,2,3]
Node (Node Leaf 1 Leaf) 2 (Node Leaf 3 Leaf)
*Tree> mktree [1,2,3,4]
mktree [1,2,3,4]
Node (Node (Node Leaf 1 Leaf) 2 Leaf) 3 (Node Leaf 4 Leaf)
*Tree> flatten $ mktree [1,2,3,4]
flatten $ mktree [1,2,3,4]
[1,2,3,4]
*Tree> fmap (*17) (mktree [1..8])
fmap (*17) (mktree [1..8])
Node (Node (Node (Node Leaf 17 Leaf) 34 Leaf) 51 (Node Leaf 68 Leaf)) 85 (Node (Node Leaf 102 Leaf) 119 (Node Leaf 136 Leaf))
*Tree> F.foldr (+) 0 (mktree [1..8])
F.foldr (+) 0 (mktree [1..8])
36
*Tree> sum [1..8]
sum [1..8]
36
*Tree> pure (*10) <*> mktree [1..8]
pure (*10) <*> mktree [1..8]
Node (Node (Node (Node Leaf 10 Leaf) 20 Leaf) 30 (Node Leaf 40 Leaf)) 50 (Node (Node Leaf 60 Leaf) 70 (Node Leaf 80 Leaf))
*Tree> (fmap (*) (mktree [1..8])) <*> (mktree [1..8])
(fmap (*) (mktree [1..8])) <*> (mktree [1..8])
Node (Node (Node (Node Leaf 1 Leaf) 4 Leaf) 9 (Node Leaf 16 Leaf)) 25 (Node (Node Leaf 36 Leaf) 49 (Node Leaf 64 Leaf))
*Tree> ((fmap (*) (mktree [1..8])) <*> (mktree [1..8]))
((fmap (*) (mktree [1..8])) <*> (mktree [1..8]))
Node (Node (Node (Node Leaf 1 Leaf) 4 Leaf) 9 (Node Leaf 16 Leaf)) 25 (Node (Node Leaf 36 Leaf) 49 (Node Leaf 64 Leaf))
*Tree> F.foldr (+) 0 ((fmap (*) (mktree [1..8])) <*> (mktree [1..8]))
F.foldr (+) 0 ((fmap (*) (mktree [1..8])) <*> (mktree [1..8]))
204
*Tree> foldr (+) 0 (flatten ((fmap (*) (mktree [1..8])) <*> (mktree [1..8])))
foldr (+) 0 (flatten ((fmap (*) (mktree [1..8])) <*> (mktree [1..8])))
204
*Tree> 